plugins/nvme/fu-plugin-nvme.c

```diff
@@ -0,0 +1,289 @@
+
+	/* we need nvme from the host system */
+	data->nvme_fn = g_find_program_in_path ("nvme");
+	if (data->nvme_fn == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "nvme is not installed");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+#if !GLIB_CHECK_VERSION(2,56,0)
+typedef GList FuObjectList;
+static void
+fu_object_list_free (FuObjectList *list)
+{
+	g_list_free_full (list, (GDestroyNotify) g_object_unref);
+}
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-function"
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(FuObjectList, fu_object_list_free)
+#pragma clang diagnostic pop
+#endif
+
+gboolean
+fu_plugin_coldplug (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+#if GLIB_CHECK_VERSION(2,56,0)
+	g_autolist(GObject) devices = NULL;
+#else
+	g_autoptr(FuObjectList) devices = NULL;
+#endif
+	devices = g_udev_client_query_by_subsystem (data->gudev_client, "nvme");
+	for (GList *l = devices; l != NULL; l = l->next) {
+		GUdevDevice *udev_device = l->data;
+		const gchar *devfile = g_udev_device_get_device_file (udev_device);
+		if (!fu_plugin_nvme_add_devfile (plugin, devfile, error))
+			return FALSE;
+	}
+	return TRUE;
+}
+
+static void
+fu_plugin_nvme_write_cb (const gchar *line, gpointer user_data)
+{
+	FuDevice *device = FU_DEVICE (user_data);
+	fu_device_set_progress (device, 1); // FIXME, add to nvme-cli
+}
+
+gboolean
+fu_plugin_update (FuPlugin *plugin,
+		  FuDevice *device,
+		  GBytes *blob_fw,
+		  FwupdInstallFlags flags,
+		  GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autofree gchar *firmware_fn = NULL;
+	g_autofree gchar *tmpdir = NULL;
+	const gchar *argv_download[] = {
+		data->nvme_fn,
+		"fw-download", fu_device_get_platform_id (device),
+		"-f", "xxx", NULL };
+	const gchar *argv_commit[] = {
+		data->nvme_fn,
+		"fw-commit", fu_device_get_platform_id (device),
+		"--slot=0", "--action=2", NULL };
+
+	/* write blob to temp location */
+	tmpdir = g_dir_make_tmp ("fwupd-XXXXXX", error);
+	if (tmpdir == NULL)
+		return FALSE;
+	firmware_fn = g_build_filename (tmpdir, "nvme-firmware.bin", NULL);
+	if (!fu_common_set_contents_bytes (firmware_fn, blob_fw, error))
+		return FALSE;
+
+	/* use nvme to write image */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	fu_device_set_progress (device, 0);
+	argv_download[4] = firmware_fn;
+	if (!fu_common_spawn_sync ((const gchar * const *) argv_download,
+				   fu_plugin_nvme_write_cb, device,
+				   NULL, error)) {
+		g_prefix_error (error, "failed to download firmware: ");
+		return FALSE;
+	}
+	if (!fu_common_spawn_sync ((const gchar * const *) argv_commit,
+				   fu_plugin_nvme_write_cb, device,
+				   NULL, error)) {
+		g_prefix_error (error, "failed to download firmware: ");
+		return FALSE;
+	}
+	fu_device_set_progress (device, 100);
+
+	/* delete temp location */
+	if (!fu_common_rmtree (tmpdir, error))
```

Comment:
I think you'll want to do this with a device locker instead to make sure that temp directory gets cleared even if the flash process failed.